### PR TITLE
[SYCL] Add max work-group size kernel properties

### PR DESCRIPTION
**Pull request by frasercrmck**
This patch adds two kernel properties to allow users to specify the maximum work-group size that a kernel will be invoked with.

The `max_work_group_size` property corresponds to the `intel::max_work_group_size` function attribute, but can be specified with 1, 2, or 3 dimensions (unlike the attribute which accepts only 3).

The `max_linear_work_group_size` property is similar but is always a single value which denotes the combined linear (total) work-group size. This can be used when the user cannot guarantee a maximum bound in each of the dimensions they wish to run the kernel, but can guarantee a total. This acts similarly to CUDA's `maxThreadsPerBlock` launch bounds property.

This patch also wires up the 'max_work_group_size' property to the equivalent SPIR-V execution mode, which should hopefully improve certain use cases.

### CODE REVIEW

**Comment by frasercrmck** [comment]
Split from #14448, @steffenlarsen and @gmlueck

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc [hunk 1]
```diff
@@ -211,6 +235,20 @@ template <> struct is_property_key<device_has_key> : std::true_type {};
...
+specified size. The number of template arguments in the `Dims` parameter pack
+must match the dimensionality of the work-group used to invoke the kernel. The
+order of the template arguments matches the constructor of the `range` class.
+
+|`max_total_work_group_size`
+|The `max_total_work_group_size` property provides a promise to the compiler
+that the kernel will never be launched with a work-group with a larger combined
+size the specified amount. The combined work-group size of work-items is
+determined as the multiplicative product of the work-group size across all
+dimensions of the work-group.
```
Comment (Pennycook) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:250]: This comment applies to the whole extension, but is best demonstrated here.

I think it would be better to use "linear" than "total" here.  The linearization rules for multi-dimensional properties in SYCL are already well-defined in other parts of the specification, and the concept of a work-group having a linear size is already present too (because that's what `group::get_local_linear_range()` returns).

You could say something like:

```suggestion
|`max_linear_work_group_size`
|The `max_linear_work_group_size` property provides a promise to the compiler
that the kernel will never be launched with a work-group with a linear size
greater than the specified amount. The linear work-group size is determined
as the multiplicative product of the work-group size across all dimensions
of the work-group.
```

Alternatively, it might even be possible to simplify this further, by saying something like:

```suggestion
|`max_linear_work_group_size`
|The `max_linear_work_group_size` property provides a promise to the compiler
that the kernel will never be launched with a work-group size where the
return value of `group::get_local_linear_range()` exceeds the specified amount.
```

Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:250]: Oh I like that idea, thank you for the suggestion!

Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc]: Done :+1:

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc [hunk 2]
```diff
@@ -59,6 +59,7 @@ Joe Garvey, Intel +
 Greg Lueck, Intel +
 John Pennycook, Intel +
 Roland Schulz, Intel
```
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:61]:
```suggestion
Roland Schulz, Intel +
```

**Review by gmlueck — COMMENTED** [review]
Spec changes LGTM.  Just a minor formatting thing.

**Review by gmlueck — APPROVED** [review]

**Review by premanandrao — COMMENTED** [review]
Could you please add a CodeGen lit test to exercise the changes in NVPTX.cpp?

**Comment by frasercrmck** [comment]
> Could you please add a CodeGen lit test to exercise the changes in NVPTX.cpp?

Sorry, do you mean a backend CodeGen LIT test, or a frontend CodeGen test?

There are separate (upstream) backend CodeGen tests already for the NVVM annotations, like [this one](https://github.com/intel/llvm/blob/sycl/llvm/test/CodeGen/NVPTX/annotations.ll).

This PR adds frontend "CodeGen" tests in [sycl/test/check_device_code/extensions/properties](https://github.com/intel/llvm/pull/14518/files#diff-27746a73b510f673f3bf664875f7a2b3941c94865260522ef4d47b05f7439c5a), as that's where the LLVM IR codegen tests for the other properties takes place. I think this is the same discussion as in #14502.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc [hunk 3]
```diff
@@ -138,6 +143,17 @@ struct work_group_size_hint_key {
   using value_t = property_value<work_group_size_hint_key, std::integral_constant<size_t, Dims>...>;
 }; // work_group_size_hint_key
 
+// Corresponds to max_work_group_size
```
Comment (AlexeySachkov) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:146]: It is not clear which `max_work_group_size` is meant here. There is a device info query like this in the core SYCL 2020 spec and there is `intel::max_work_group_size` attribute which is not covered by any SYCL extension specifications AFAIK. Do we really need this comment?

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc [hunk 4]
```diff
@@ -211,6 +235,18 @@ template <> struct is_property_key<device_has_key> : std::true_type {};
  of the work-group used to invoke the kernel. The order of the template
  arguments matches the constructor of the `range` class.
 
+|`max_work_group_size`
+|The `max_work_group_size` property provides a promise to the compiler
+that the kernel will never be launched with a larger work-group than the
+specified size. The number of template arguments in the `Dims` parameter pack
+must match the dimensionality of the work-group used to invoke the kernel. The
+order of the template arguments matches the constructor of the `range` class.
```
Comment (AlexeySachkov) [inline sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:243]: Should we document which exception is thrown when runtime-provided work-group size is bigger than one specified by the property? The same applies to the other property being added.

File: clang/lib/CodeGen/Targets/NVPTX.cpp [hunk 5]
```diff
@@ -301,6 +324,33 @@ void NVPTXTargetCodeGenInfo::setTargetAttributes(
...
+        for (auto [AnnotStr, Val] : zip(Annots, Ops))
+          if (Val.has_value())
+            addNVVMMetadata(F, AnnotStr, *Val);
+      }
+    }
+
+    if (auto Attr = F->getFnAttribute("sycl-max-linear-work-group-size");
+        Attr.isValid()) {
+      size_t Value = 0;
+      bool Error = Attr.getValueAsString().getAsInteger(10, Value);
```
Comment (AlexeySachkov) [inline clang/lib/CodeGen/Targets/NVPTX.cpp:349]: `[[maybe_unused]]` is missing here

**Review by KseniyaTikhomirova — APPROVED** [review]
SYCL part looks ok to me.

**Review by AlexeySachkov — APPROVED** [review]
`CompileTimePropertiesPass` changes LGTM

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc (hunk 4 above)
Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:243]: Yes, we probably should. I think we just want to copy the exception behaviour from `reqd_work_group_size`:

``` quote
If the kernel is submitted to a device that does support the work-group size,
but the application provides an nd_range that does not match the size from
the attribute, then the implementation must throw a synchronous exception with the
errc::nd_range error code.
```

Unfortunately it doesn't seem like we're actually doing this, at least when compiling SYCL for CUDA - even for the attribute.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc (hunk 3 above)
Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:146]: No I don't think so. As you say, it's an unspecified attribute so isn't helpful to point to it.

File: clang/lib/CodeGen/Targets/NVPTX.cpp (hunk 5 above)
Comment (frasercrmck) [reply clang/lib/CodeGen/Targets/NVPTX.cpp:349]: Good spot, thanks!

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc (hunk 4 above)
Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: I've added wording to this effect. @gmlueck do you want to take another look?

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: The specification looks good, but you say in the comment above that the implementation doesn't actually do this.  Can you add that support?

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: Oh ... we document that these kernel properties work for all backends.  Do we have e-2-e tests that are enabled for all backends?  Those tests should also check for the exception (as we do in the test you reference above).

I see what you mean about the existing problem that CUDA and HIP don't generate the exception.  Do we have a tracker that captures this?  In general, we shouldn't disable tests unless we have a tracker that captures the work to fix the problem.

Can we:

* Write tests for these new properties, including tests that check the exception
* Disable the exception tests on CUDA / HIP with a comment that this isn't supported yet
* Make sure we have a tracker to implement the exception
* Make time to resolve this tracker

Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: Yeah, seems like this issue has flown under the radar for a while - we weren't aware of the problem. I've created an internal ticket to fix the current bug for HIP/CUDA, and that work will presumably also fix these new properties too.

Regarding end-to-end testing for exceptions, I fear we might have backed ourselves into a corner:

* HIP/CUDA targets are broken for the aforementioned reasons
* The OpenCL CPU target doesn't support the `KernelAttributesINTEL` SPIR-V execution mode, so doesn't even support the `[[intel::max_work_group_size]]` _attribute_, let alone the kernel property.
* The OpenCL FPGA target does support that execution mode, but doesn't throw an exception, even for the attribute. Note how there are no end-to-end tests for the attribute either! This is arguably a bug with that target, but given that there's no specification for the attribute, it's not a firm footing. I think UR would need to receive this information from the Level Zero driver, and so we'd need to update the Level Zero specification to be able to provide those kernel properties.
* I haven't been able to test any other targets.

We would probably also need to extend `SPV_INTEL_kernel_attributes` to support the `max_linear_work_group_size` execution mode. Otherwise the information is lost. But then again the OpenCL CPU target doesn't support that extension.

If we do want to have these attributes in a specification, we definitely want to specify that they can throw an exception - otherwise they're useless. I don't think "can" is good enough either: we want guarantees. It just so happens that we're unable to meaningfully test this behaviour on any target right now.

Ironically it seems like HIP/CUDA might be the easiest to get working as they don't use SPIR-V.

Do you have any suggestions on how to proceed? It's an experimental extension but I don't know if that means we can half implement it. Can we selectively implement it for different targets? If the OpenCL CPU driver is never going to support the required SPIR-V mode, there's not much we can do.

Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: Note that #14634 re-enables those tests for HIP/CUDA so the fundamental issue should be fixed.

**Comment by frasercrmck** [comment]
> Could you please add a CodeGen lit test to exercise the changes in NVPTX.cpp?

Note the updated version of this patch doesn't touch `clang` any more once #14634 landed.

File: sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc (hunk 4 above)
Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: @gmlueck how should we proceed with this issue? It's my opinion that this extension can't be supported by all targets (Intel targets), at least not in this PR. There are far deeper issues going on with both of those targets, least of all no support in SPIR-V for one of the properties.

We can make NVPTX/AMD support the property, but it'd involve a separate change to UR and for us to bump UR. Should that be part of this PR, or done as a follow-up?

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: Sorry, can you summarize the current state on the various backends?  Are these two new properties supported on *any* backend currently?

Also, two questions regarding support on Intel devices via SPV:

* You mention above that the OpenCL CPU backend doesn't support the `KernelAttributesINTEL` execution mode.  What happens if you submit a module to that backend with this attribute?  Is it ignored, or does it raise an error?

* What is the support on the Level Zero and OpenCL GPU backends?

Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: > Sorry, can you summarize the current state on the various backends? Are these two new properties supported on _any_ backend currently?
> 
> Also, two questions regarding support on Intel devices via SPV:
> 
>     * You mention above that the OpenCL CPU backend doesn't support the `KernelAttributesINTEL` execution mode.  What happens if you submit a module to that backend with this attribute?  Is it ignored, or does it raise an error?
> 
>     * What is the support on the Level Zero and OpenCL GPU backends?

These properties are not fully supported on any backend, currently. The CUDA backend uses it for what it needs it for - to emit NVVM annotations - but no backend supports the exceptions:

CUDA, HIP, and Native CPU would need the SYCL runtime to query the UR adapters to receive the max work-group size information, and throw the right exception. This would require an update to the UR spec.

OpenCL CPU reports:
```
# | Build program log for '11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz':
# | Compilation started
# | Unsupported SPIR-V module
# | KernelAttributes is only supported on FPGA emulator
# | Compilation failed
```
and crashes.

I don't really know how OpenCL GPU works. Presumably it uses SPIR-V, but would also need the change to UR to query for the program information. Note also that SPIR-V can only encode `MaxWorkGroupSize` but not "MaxLinearWorkGroupSize" so we'd need a new SPIR-V extension, or to update the `KernelAttributesINTEL` extension, before we can report an exception for SPIR-V programs. Either way, all of our SPIR-V consumers would need updating.

I also don't fully understand the Level Zero backend either. If it uses SPIR-V it will need the same spec changes as for OpenCL GPU. Regardless it would need the SYCL runtime to query UR (as before) which would then need to query Level Zero for the program information. This would require a change to the L0 specification.

I think that summarizes the current situation. Adhering to the exception behaviour would require at least one specification change for every backend (UR being essential for all of them), and I don't see an easy way for the OpenCL CPU given its behaviour with the required SPIR-V extension, and the lack of SPIR-V/LevelZero support for the "linear" property.

Maybe it's a chicken and egg situation, because to add these properties to UR we'd want to be reasonably assured that all backends _can_ support them, at least in principle.

Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: @gmlueck I've preemptively added these two properties to UR in https://github.com/oneapi-src/unified-runtime/pull/1996, and supported them for CUDA and Native CPU. The other adapters are marked as XFAIL, pending changes to SPIR-V/OpenCL/Level Zero specs. That might be enough to get this PR through the door?

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: Let's do this.  Add a new section to the specification titled "Kernel Properties for the CUDA backend", or you can title it "Kernel Properties for the CUDA and native CPU backends" if you think the native cpu support is interesting.  This section should immediately follow "Kernel properties", and it should have an introduction that says something like:

> The kernel properties specified in this section may only be used to decorate kernels that are submitted to the CUDA backend.  Attempting to submit a kernel with these properties to another backend has undefined behavior.

Then move the synopsis and description of these two properties to that new section.

Do not include the paragraph saying that an exception is raised since that is not the case.  However, do add a note like this:

> [_Note:_ In the future, implementations may throw a synchronous exception with the `errc::nd_range` error code if the kernel is submitted with an `nd_range` that exceeds the size specified by the property.  However, that is not currently the case.  Instead, the behavior is undefined if the size of the `nd_range` exceeds the size of the property.
_{endnote}_]

(See [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_memory.asciidoc) for an example of this "note" syntax.)

Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: Okay, that sounds good. But I should say (I maybe didn't make this clear) that the exception *is* currently raised for CUDA and Native CPU. It could be enabled for HIP with minimal effort, but we're currently not prioritising that adapter. If that would help come up with a better rationale than "Kernel Properties for the CUDA and Native CPU and HIP backends" then it might be worthwhile doing the work for HIP too.

I don't know if that changes your suggestion about moving these properties to a new section or not, but I would imagine we do want to keep the wording on exception behaviour.

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc:242]: OK, I didn't realize that CUDA and native CPU implemented the exception.  In that case, we should still create the separate section, and also keep the wording about the exception.  The "note" is not needed.

Comment (frasercrmck) [reply sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc]: Great, thank you. That's done now.

**Comment by frasercrmck** [comment]
ping @intel/dpcpp-spirv-reviewers, thanks

**Comment by frasercrmck** [comment]
ping @intel/dpcpp-spirv-reviewers

File: llvm-spirv/lib/SPIRV/PreprocessMetadata.cpp [hunk 6]
```diff
@@ -188,16 +188,17 @@ void PreprocessMetadataBase::visit(Module *M) {
     //         i32 Y, i32 Z}
     if (MDNode *MaxWorkgroupSizeINTEL =
             Kernel.getMetadata(kSPIR2MD::MaxWGSize)) {
-      assert(MaxWorkgroupSizeINTEL->getNumOperands() == 3 &&
-             "max_work_group_size does not have 3 operands.");
+      assert(MaxWorkgroupSizeINTEL->getNumOperands() >= 1 &&
```
Comment (AlexeySachkov) [inline llvm-spirv/lib/SPIRV/PreprocessMetadata.cpp:191]: Our version of `llvm-spirv` is considered to be read-only, or at least we strive to have it as such. Unless there is a strong justification, any changes to the translator should first go into the upstream project and then be cherry-picked or brought by a regular pulldown

Comment (frasercrmck) [reply llvm-spirv/lib/SPIRV/PreprocessMetadata.cpp:191]: Ah right, I wasn't aware - thanks. I've created https://github.com/KhronosGroup/SPIRV-LLVM-Translator/pull/2711

Comment (frasercrmck) [reply llvm-spirv/lib/SPIRV/PreprocessMetadata.cpp]: This was merged (thanks) and now pulled down, so we can resolve this.

File: sycl/cmake/modules/FetchUnifiedRuntime.cmake [hunk 7]
```diff
@@ -116,14 +116,15 @@ if(SYCL_UR_USE_FETCH_CONTENT)
...
-  set(UNIFIED_RUNTIME_TAG 7a2caca559da81620c7430521873045bf42eafc1)
+  set(UNIFIED_RUNTIME_REPO "https://github.com/frasercrmck/unified-runtime.git")
+    
+  # commit cabf128094eff9ff7b79bdff559640a8a111f0c3
+  # Merge: a96fcbc5 15bca3b6
+  # Author: Omar Ahmed <omar.ahmed@codeplay.com>
+  # Date:   Mon Aug 19 16:20:45 2024 +0100
+  #     Merge pull request #1984 from rafbiels/rafbiels/cuda-stream-race-cond
+  #     Fix race condition in CUDA stream creation
+  set(UNIFIED_RUNTIME_TAG 6eb5208bad81457d367dbb5b51ad266cda7cf7c1)
```
Comment (aarongreig) [inline sycl/cmake/modules/FetchUnifiedRuntime.cmake:127]: https://github.com/oneapi-src/unified-runtime/pull/1996 is merged, please update

Comment (frasercrmck) [reply sycl/cmake/modules/FetchUnifiedRuntime.cmake]: done, thanks!

**Review by aarongreig — APPROVED** [review]

**Comment by frasercrmck** [comment]
Three E2E tests failing on linux:

```
Failed Tests (3):
  SYCL :: DeviceDependencies/free_function_kernels.cpp
  SYCL :: LLVMIntrinsicLowering/bitreverse.cpp
  SYCL :: LLVMIntrinsicLowering/sub_byte_bitreverse.cpp
```

All for the same reason:
```
clang++: error: option '-fno-sycl-allow-device-dependencies' is deprecated and will be removed in a future release
```

Looks like a pre-existing problem.

**Comment by frasercrmck** [comment]
Same on Windows:

```
********************
Failed Tests (2):
  SYCL :: DeviceDependencies/free_function_kernels.cpp
  SYCL :: LLVMIntrinsicLowering/bitreverse.cpp
```

**Comment by frasercrmck** [comment]
@intel/llvm-gatekeepers this is ready to merge, thanks. The failures are unrelated (see above).
